### PR TITLE
Prefer range based for loops.

### PR DESCRIPTION
**Pull request by abarret**
This is a run of modernize-loop-convert with clang-tidy.

I'm still working on removing unused `typedef`'s, but I went ahead and opened this so changes can be reviewed.

### CODE REVIEW

**Comment by boyceg** [comment]
@abarret is this ready for review?

File: ibtk/include/ibtk/private/LDataManager-inl.h [hunk 1]
```diff
@@ -165,11 +165,9 @@ LDataManager::getLagrangianStructureNames(const int level_number) const
     TBOX_ASSERT(d_coarsest_ln <= level_number && d_finest_ln >= level_number);
 #endif
     std::vector<std::string> ret_val;
-    for (std::map<int, std::string>::const_iterator cit(d_strct_id_to_strct_name_map[level_number].begin());
-         cit != d_strct_id_to_strct_name_map[level_number].end();
-         ++cit)
+    for (const auto& map : d_strct_id_to_strct_name_map[level_number])
```
Comment (boyceg) [inline ibtk/include/ibtk/private/LDataManager-inl.h]: 👍

File: ibtk/src/boundary/physical_boundary/muParserRobinBcCoefs.cpp [hunk 2]
```diff
@@ -244,12 +244,12 @@ muParserRobinBcCoefs::muParserRobinBcCoefs(const std::string& object_name,
     const double pi = 3.1415926535897932384626433832795;
     const double* const xLower = grid_geom->getXLower();
     const double* const xUpper = grid_geom->getXUpper();
-    for (auto cit = all_parsers.begin(); cit != all_parsers.end(); ++cit)
+    for (auto& all_parser : all_parsers)
```
Comment (boyceg) [inline ibtk/src/boundary/physical_boundary/muParserRobinBcCoefs.cpp]: `parser`

File: ibtk/src/lagrangian/FEDataInterpolation.cpp [hunk 3]
```diff
@@ -225,9 +225,9 @@ FEDataInterpolation::init(const bool use_IB_ghosted_vecs)
         const System& system = *d_systems[system_idx];
         const DofMap& system_dof_map = system.get_dof_map();
         const std::vector<int>& all_vars = d_system_all_vars[system_idx];
-        for (unsigned int k = 0; k < all_vars.size(); ++k)
+        for (const auto& all_var : all_vars)
```
Comment (boyceg) [inline ibtk/src/lagrangian/FEDataInterpolation.cpp]: `var`

Comment (drwells) [inline ibtk/src/lagrangian/FEDataInterpolation.cpp]: (same here, with all_var)

**Comment by abarret** [comment]
I updated the variable names as well as a few `it` and `cit` names I missed. If there's nothing else, I can squash into a single commit.

**Comment by boyceg** [comment]
@drwells, should we prefer to use `const auto&` in range-based for loops? If so, @abarret, could you switch everything to initialize using `const auto&` and then back off to `auto&` or `auto` where needed? (Perhaps you already have done that.)

Looks great!

**Comment by drwells** [comment]
Yes.

**Comment by abarret** [comment]
Done. I used the compiler to check for correctness on the changes. Is there any case the compiler wouldn't catch for this kind of switch?

**Comment by boyceg** [comment]
@abarret, do you mean, are there cases where `const auto&` is not the right thing to use in the range-based for loop? The only cases are where the loop is actually modifying the objects that are being iterated over.

The following is not OK:
```
vector<T> vec = ... ;
for (const auto& v : vec)
{
    v.non_const_member_function();
}
```
but the following is OK:
```
vector<T*> vec = ... ;
for (const auto& v : vec)
{
    v->non_const_member_function();
}
```
because here, `v` is a constant reference to a `T*`.

On the other hand, this does not work:
```
vector<const T*> vec = ... ;
for (const auto& v : vec)
{
    v->non_const_member_function();
}
```
The reason is that here, `v` is a const reference to a `const T*`, so we cannot call any non-const member functions.

I am 99% sure that if you default to using `const auto&` in the range-based for loops, then basically all you will need to do is fix any cases where the compiler complains about const correctness.

**Comment by boyceg** [comment]
In cases where the compiler does complain about const correctness, you will need to determine whether you want a copy (`auto`) or a reference (`auto&`) or, possibly, an r-value reference (`auto&&`).

**Comment by abarret** [comment]
> I am 99% sure that if you default to using `const auto&` in the range-based for loops, then basically all you will need to do is fix any cases where the compiler complains about const correctness.

This is what I did. I couldn't think of a case where the compiler wouldn't catch an error, but I also can't think of many things, so I wanted to double check.

I went back and checked all the range-based for loops that didn't use `const auto&` and I don't think any of those need to be changed. I'm done making changes unless someone else (@drwells?) suggests changes.

File: src/IB/IBLagrangianForceStrategySet.cpp [hunk 4]
```diff
@@ -61,9 +61,9 @@ IBLagrangianForceStrategySet::~IBLagrangianForceStrategySet()
 void
 IBLagrangianForceStrategySet::setTimeInterval(const double current_time, const double new_time)
 {
-    for (auto cit = d_strategy_set.begin(); cit != d_strategy_set.end(); ++cit)
+    for (const auto& set : d_strategy_set)
```
Comment (boyceg) [inline src/IB/IBLagrangianForceStrategySet.cpp]:
```suggestion
    for (const auto& strategy : d_strategy_set)
```

File: src/IB/IBLagrangianForceStrategySet.cpp [hunk 5]
```diff
@@ -61,9 +61,9 @@ IBLagrangianForceStrategySet::~IBLagrangianForceStrategySet()
...
     {
-        (*cit)->setTimeInterval(current_time, new_time);
+        set->setTimeInterval(current_time, new_time);
```
Comment (boyceg) [inline src/IB/IBLagrangianForceStrategySet.cpp]:
```suggestion
        strategy->setTimeInterval(current_time, new_time);
```

**Comment by abarret** [comment]
@boyceg I've fixed the naming issues. Once you give me the thumbs up on the rest of the changes I will squash and fix the conflicts. I'm pretty sure these conflicts are just overlaps with #367.

File: ibtk/src/lagrangian/LSiloDataWriter.cpp [hunk 6]
```diff
@@ -469,9 +468,9 @@ build_local_ucd_mesh(DBfile* dbfile,
 
     // Prune duplicate edges.
     std::set<std::pair<int, int> > local_edge_set;
-    for (auto it = edge_map.begin(); it != edge_map.end(); ++it)
+    for (const auto& map : edge_map)
```
Comment (boyceg) [inline ibtk/src/lagrangian/LSiloDataWriter.cpp]: Is this a map?

File: src/IB/IBRedundantInitializer.cpp [hunk 7]
```diff
@@ -403,9 +403,9 @@ IBRedundantInitializer::initializeSprings()
 
                 int min_idx = 0;
                 int max_idx = d_num_vertex[ln][j];
-                for (auto it = d_spring_edge_map[ln][j].begin(); it != d_spring_edge_map[ln][j].end(); ++it)
+                for (const auto& edge_map : d_spring_edge_map[ln][j])
```
Comment (boyceg) [inline src/IB/IBRedundantInitializer.cpp]: Are these maps? Or edges?

**Comment by boyceg** [comment]
Couple more tiny things that I just commented on. Also, I think that most occurrences of
```
for (auto& k : is_XXX)
```
should be
```
for (auto& is : is_XXX)
```

It might be worth doing a quick check for range-based for loops where the variable name is a common index variable name (`d`, `i`, `j`, `k`, `n` come to mind) and see if any of those now no longer refer to indexes but to actual variables.

After you do that, feel free to squash, and I will merge once Jenkins give the 👍.

File: ibtk/src/lagrangian/LSiloDataWriter.cpp (hunk 6 above)
Comment (abarret) [reply ibtk/src/lagrangian/LSiloDataWriter.cpp]: I'm not sure what an appropriate name would be. `edge_map` is of type `std::multimap<int, std::pair<int, int> >`. Perhaps `elem` for an element of the map? This kind of situation happens several times

File: src/IB/IBRedundantInitializer.cpp (hunk 7 above)
Comment (drwells) [reply src/IB/IBRedundantInitializer.cpp]: They are `std::pair<int, Edge>`s. I usually write something like `edge_pair`.

**Comment by boyceg** [comment]
Or just edge (or even just e for edge).

**Comment by abarret** [comment]
@boyceg I fixed a few more names. This branch is now squashed and rebased on top of master

**Comment by boyceg** [comment]
Let's do this.

**Pull request merged**